### PR TITLE
feat(mv3-part-2): Clean up BrowserAdapterEventManager

### PR DESCRIPTION
**Pull request by madalynrose**
#### Details

This addresses outstanding comments from @dbjorge on #5380 that weren't seen before it got merged.

This PR:
* removes AlarmUtils as they are unnecessary as long as we ensure setTimeout can finish
* updates the BrowserAdapterEventManager to register a single listener per eventType
* updates logic around event deferral to create a promise to be resolved externally, return pending promise, and resolve promise when the deferred event is handled by an application listener
* adds a function to promiseFactory for creating externally resolved promises

##### Motivation

feature work

##### Context

None of this code is called anywhere yet but will be used by the browser adapter used in the service worker.

Note that since we're tracking one application listener per eventType, we will need to create a single application listener to react to and dispatch events for eventTypes that have multiple calls to addListener (e.g. onMessage will need to respond to different types of messages).

#### Pull request checklist

- [n/a] Addresses an existing issue: #0000
- [x] Ran `yarn fastpass`
- [x] Added/updated relevant unit test(s) (and ran `yarn test`)
- [x] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [x] PR title *AND* final merge commit title both start with a semantic tag (`fix:`, `chore:`, `feat(feature-name):`, `refactor:`). See `CONTRIBUTING.md`.

### CODE REVIEW

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:41]: Maybe have this be defensive about multiple registration (ie, detect and throw, rather than silently overriding)?

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:32]: I think this needs to pass through the return value of `this.processEvent`

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:127]: `this.adapterListener` isn't memoized, so I don't think this is going to have the desired effect

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:131]: This check isn't necessary, `delete` is idempotent in javascript.

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:88]: This block isn't quite right; it needs to be returning a promise which will resolves after two minutes. This is instead returning a promise which resolves immediately, and setting up a `setTimeout` which will create an unrelated Promise after two minutes and then immediately throw it away.

Comment (dbjorge) [reply src/common/browser-adapters/browser-adapter-event-manager.ts:88]: To make it easier to test this without actually needing to wait for timeouts, I'd suggest creating a `promiseFactory.delay` which does something like `return new Promise((resolve) => { this.setTimeout(() => resolve(), delayInMillis); });`

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:79]: I don't think rethrowing like this is quite right; I think that even if the listener throws an error, it's probably safest for us to still apply the rest of the logic and return a promise to the browser event (otherwise, a single misbehaving event handler is liable to cause the whole service worker to be torn down early).

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:85]: Do we have any cases where an application handler will be returning a non-undefined, non-promise result? I'm not sure the "delay for 2 minutes" approach would be a good idea in that case, we probably don't want to delay actual responses for 2 minutes. I don't think we have a great solution for that case - we don't have a way to simultaneously report a response back and also indicate the 2 minute deferal.

I think we should probably fail fast and log an error if that happens, and update any such cases that occur in practice by auditing them for any fire and forget behavior, updating them to ensure any such behavior is captured within a Promise, and then returning a Promise with the original result.

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:54]: Prefer using for-of loops rather than `Array.forEach` - the former gives better stack traces from `Error`s and is more straightforward to step through in a debugger

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:111]: This `find` call gives `processDeferedEvents` an `O(deferredEvents.length ^ 2)` time complexity, and we do see thousands of events in normal usage of the app. "Are we in an already-deferred" path is already known in each of the 2 `processEvent` callsites, and one of them doesn't actually use the return value of `processEvent`; maybe instead of the current contents of this else branch, just return `null` and update L32 to be something along the lines of `return processEvent(e, ...args) ?? deferEvent(e, ...args);`?

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:95]: I don't think this has the desired effect in the case where the browser sends us an event we never register a handler for. I think what we'd want in that case would be to time out the promise after 4 minutes, but I think this would just pend the promise indefinitely and get our worker torn down after 5 minutes

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:84]: `PromiseFactory.timeout` _rejects_ after the given timeout, and I'm not sure we actually want that here as opposed to resolving after the given timeout. Rejecting is more accurate, but if the browser responds to a rejected promise by killing the service worker, we probably don't want to do that. I don't know if the Browser treats resolved vs rejected promises the same way.

**Review by dbjorge — CHANGES_REQUESTED** [review]
Thanks for going back for the comments! Reviewed inline

File: src/tests/unit/tests/common/promises/promise-factory.test.ts
Comment (dbjorge) [inline src/tests/unit/tests/common/promises/promise-factory.test.ts:66]: This is missing a `testObject` invocation, no?

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:103]: I think `delay` is right for the behavior on L108, but I don't think it's quite what what we want here - if an application handler returns a `Promise` that resovles in a timely fashion, I think we should probably trust it to encapsulate the required async work and resolve as soon as it does, not delay it for 4 minutes. I think the 4 minutes should be a limit, not a minimum delay.

I think a straightforward way to achieve that but still avoid rejecting the promise would be something like

```ts
return this.promiseFactory.timeout(result, FOUR_MINUTES)
    .catch(() => {
        // We want to avoid rejecting the Promise we give back to the browser event handler
        // because doing so is liable to cause the browser to tear down our Service Worker
        // immediately, even if other events are in progress.
        return undefined;
    });
```

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:97]: Can this inject a `Logger` dependency and use `logger.error`, rather than calling `console.error` directly? Even if the message isn't important to test, doing that and using a noop logger avoids unnecessary console spew during test execution.

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:25]: It would maybe be helpful to give this class a comment explaining the problem it's solving; I worry that it would be easy for a maintainer without context to remove a subtly important timeout

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (madalynrose) [reply src/common/browser-adapters/browser-adapter-event-manager.ts:97]: good call!

File: src/tests/unit/tests/common/promises/promise-factory.test.ts
Comment (madalynrose) [reply src/tests/unit/tests/common/promises/promise-factory.test.ts:66]: 🤦🏻‍♀️ you're right! thanks!

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:43]: I'm not sure I understand the comment here; the new behavior for the 4 minute timeout promise turns a "timeout" from a event that rejects the promise to an event that resolves the Promise as undefined, but the value that `processedResults` gets set to here is still going to be a Promise, not `undefined`.

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:148]: I think it would be good for us to log an error in this case as well; I don't think this is expected in any scenario

File: src/common/browser-adapters/browser-adapter-event-manager.ts
Comment (dbjorge) [inline src/common/browser-adapters/browser-adapter-event-manager.ts:88]: I don't think it's quite right for this to ignore the return value of `this.processEvent`; I think this needs to set it up so that if `processEvent` returns a `Promise`, the `deferredResolution.resolveHook` gets invoked only after that promise settles, not immediately.

**Review by dbjorge — DISMISSED** [review]
Thanks for the updates! I've gone back through and marked resolved all the original comments. Left another few (the one about the `processEvents`/`deferredResolution` relationship is blocking, I think)

**Comment by dbjorge** [comment]
I've cleaned up the outstanding issues. I also ended up redoing the unit test strategy; I created simulators for the browser events and underlying promise factory to avoid needing to test implementation details.

I would recommend the next reviewer review the changes as a whole, rather than just the delta since Madalyn's last commit - this new behavior is complex enough that I'd like a second pair of eyes on the whole thing anyway.

**Review by sfoslund — COMMENTED** [review]
Looks good overall, a few nits/ questions!

File: src/common/browser-adapters/browser-event-manager.ts
Comment (dbjorge) [reply src/common/browser-adapters/browser-event-manager.ts:104]: I think it's preferable to avoid processing an already-timed-out event. If a still-deferred event ever becomes stale, it means that it timed out while it was still in a deferred state, and we've already indicated back to the browser that this event should no longer be preventing the worker from being torn down. If that were to happen, I expect debugging the resulting timeout error log on its own will probably be less misleading than simultaneously debugging a timeout error log plus also whatever exciting result happens when the worker gets torn down mid-handler-execution.

File: src/common/browser-adapters/browser-event-manager.ts
Comment (dbjorge) [reply src/common/browser-adapters/browser-event-manager.ts:31]: Excellent question. Added a link to the relevant Chromium docs to the class-level comment

File: src/common/browser-adapters/browser-event-manager.ts
Comment (dbjorge) [reply src/common/browser-adapters/browser-event-manager.ts:107]: (added a comment to this effect)

**Review by sfoslund — APPROVED** [review]
Looks great! I appreciate all the detailed comments!

**Pull request merged**
